programs/ziti-edge-tunnel/windows-scripts.c

```diff
@@ -0,0 +1,355 @@
+    struct modify_service_nrpt_req *modify_svc_req_data = ar->data;
+    uv_loop_t *nrpt_loop = ar->loop;
+
+    struct add_service_nrpt_req *add_svc_req_data = calloc(1, sizeof(struct add_service_nrpt_req));
+
+    uv_close((uv_handle_t *) ar, (uv_close_cb) free);
+
+    remove_nrpt_rules_script(nrpt_loop, modify_svc_req_data->hostnamesToRemove);
+    add_nrpt_rules_script(nrpt_loop, add_svc_req_data);
+
+}
+
+void remove_single_nrpt_rule(char* nrpt_rule) {
+    char remove_cmd[MAX_POWERSHELL_COMMAND_LEN];
+    size_t buf_len = sprintf(remove_cmd, "powershell -Command \"Get-DnsClientNrptRule | where Namespace -eq '%s' | Remove-DnsClientNrptRule -Force -ErrorAction SilentlyContinue\"", nrpt_rule);
+    ZITI_LOG(TRACE, "Removing nrpt rule. total script size: %d", buf_len);
+
+    ZITI_LOG(INFO, "Executing Remove nrpt rule: %s", remove_cmd);
+    int rc = system(remove_cmd);
+    if (rc != 0) {
+        ZITI_LOG(WARN, "Delete single NRPT rule: %d(err=%d)", rc, GetLastError());
+    }
+}
+
+bool is_nrpt_policies_effective(char* tns_ip) {
+    char add_cmd[MAX_POWERSHELL_COMMAND_LEN];
+    size_t buf_len = sprintf(add_cmd, "powershell -Command \"Add-DnsClientNrptRule -Namespace '.ziti.test' -NameServers '%s' -Comment 'Added by ziti-tunnel' -DisplayName 'ziti-tunnel:.ziti.test'\"",tns_ip);
+    ZITI_LOG(TRACE, "add test nrpt rule. total script size: %d", buf_len);
+
+    ZITI_LOG(INFO, "Executing add test nrpt rule. %s", add_cmd);
+    int rc = system(add_cmd);
+    if (rc != 0) {
+        ZITI_LOG(WARN, "Add test NRPT rule: %d(err=%d)", rc, GetLastError());
+        return false;
+    }
+
+    char get_cmd[MAX_POWERSHELL_COMMAND_LEN] = "powershell -Command \"Get-DnsClientNrptPolicy -Effective | Select-Object Namespace -Unique | Where-Object Namespace -Eq '.ziti.test'\"";
+    char* result = exec_process_fetch_result(get_cmd);
+    if (result == NULL) {
+        ZITI_LOG(WARN, "get test nrpt rule script failed");
+        return false;
+    } else {
+        char delim[] = "\r\n";
+        char *token;
+        boolean policy_found = false;
+
+        token = strtok(result, delim);
+        while( token != NULL ) {
+            if (strcmp(token, ".ziti.test") == 0) {
+                policy_found = true;
+                break;
+            }
+            token = strtok(NULL, delim);
+        }
+        free(result);
+
+        if (policy_found) {
+            ZITI_LOG(INFO, "NRPT policies are effective in this system");
+            remove_single_nrpt_rule(".ziti.test");
+            return true;
+        } else {
+            ZITI_LOG(INFO, "NRPT policies are ineffective in this system");
+            return false;
+        }
+
+    }
+}
+
+model_map *get_connection_specific_domains() {
+    char get_cmd[MAX_POWERSHELL_COMMAND_LEN] = "powershell -Command \"Get-DnsClient | Select-Object ConnectionSpecificSuffix -Unique | ForEach-Object { $_.ConnectionSpecificSuffix }; (Get-DnsClientGlobalSetting).SuffixSearchList\"";
```

Comment:
you might as well declare this as a `const char *`, since the compiler will know exactly how much storage it needs. declaring as an array means that the buffer will either be too large (wasted memory) or too small (potential overrun).